http-server-netty/src/main/java/io/micronaut/http/server/netty/shortcircuit/DiscriminatorStage.java

```diff
@@ -0,0 +1,144 @@
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * The different stages of route matching. These are matched in order. For example the
+ * {@link #PATH} stage matches the request path to the candidate routes.<br>
+ * This class also handles transforming the {@link MatchRule}s corresponding to each stage to a
+ * {@link MatchPlan}.
+ *
+ * @author Jonas Konrad
+ * @since 4.3.0
+ */
+@Internal
+enum DiscriminatorStage {
+    PATH {
+        @Override
+        <R> MatchPlan<R> planDiscriminate(Map<MatchRule, MatchPlan<R>> nextPlans) {
+            // simplify already removed any overlaps between patterns and exact matches. only the exact matches are left
+            Map<String, MatchPlan<R>> byString = nextPlans.entrySet().stream()
+                .filter(e -> e.getKey() instanceof MatchRule.PathMatchExact)
+                .collect(Collectors.toMap(e -> ((MatchRule.PathMatchExact) e.getKey()).path(), Map.Entry::getValue));
+            return request -> {
+                // this replicates AbstractNettyHttpRequest.getPath but not exactly :(
+                URI uri;
+                try {
+                    uri = URI.create(request.uri());
+                } catch (IllegalArgumentException iae) {
+                    return ExecutionLeaf.indeterminate();
+                }
+                String rawPath = new QueryStringDecoder(uri).rawPath();
+                if (!rawPath.isEmpty() && rawPath.charAt(rawPath.length() - 1) == '/') {
+                    rawPath = rawPath.substring(0, rawPath.length() - 1);
+                }
+                MatchPlan<R> plan = byString.get(rawPath);
+                return plan == null ? ExecutionLeaf.indeterminate() : plan.execute(request);
+            };
+        }
+    },
+    METHOD {
+        @Override
+        <R> MatchPlan<R> planDiscriminate(Map<MatchRule, MatchPlan<R>> nextPlans) {
+            Map<HttpMethod, MatchPlan<R>> byMethod = coerceRules(MatchRule.Method.class, nextPlans)
+                .entrySet().stream().collect(Collectors.toMap(e -> HttpMethod.valueOf(e.getKey().method().name()), Map.Entry::getValue));
+            return request -> {
+                MatchPlan<R> plan = byMethod.get(request.method());
+                return plan == null ? ExecutionLeaf.indeterminate() : plan.execute(request);
+            };
+        }
+    },
+    CONTENT_TYPE {
+        @Override
+        <R> MatchPlan<R> planDiscriminate(Map<MatchRule, MatchPlan<R>> nextPlans) {
+            Map<String, MatchPlan<R>> byContentType = coerceRules(MatchRule.ContentType.class, nextPlans)
+                .entrySet().stream().collect(Collectors.toMap(e -> {
+                    MediaType expectedType = e.getKey().expectedType();
+                    return expectedType == null ? null : expectedType.getName();
```

Comment:
a null key?

Reply:
For MatchRule.ContentType, `null` means no content type header. it's convenient because `request.headers().get(HttpHeaderNames.CONTENT_TYPE)` returns `null` when the content type is missing, which will then match `MatchRule.ContentType(null)`